### PR TITLE
[CYS] Patterns Migration: Update the registration of patterns within WooCommerce, ensuring they are fetched from the relevant PTK categories

### LINKED ISSUE #47396: CYS > Patterns Migration: Update the registration of patterns within WooCommerce, ensuring they are fetched from the relevant PTK categories
**Issue Description:**

Update the registration of patterns within WooCommerce, ensuring they are fetched from the relevant PTK categories. We should still keep a fallback for local registration as a few patterns will be preserved within the plugin in case of connectivity issues and/or if the user didn't allow usage tracking.

More details about implementation are available on: pdnLyh-69o-p2

Designs: FlPl7fHaGv9mpu8Nfl0joH-fi-2814_125908

The following pattern categories should be fetched from WP.com's PTK:

* Intro 
* About
* Services
* Testimonials

### PR DESCRIPTION
**Pull request by albarin**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/47396

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

The main feature this PR adds is making available patterns coming from the PTK API that belong to the following categories: `Intro`, `About`, `Services`, and `Testimonials` (only if the user allows tracking).
Note: the feature is under the `pattern-toolkit-full-composability` feature flag.

This PR also includes a refactor of the `BlockPatterns` class. This is an explanation of the changes and classes created:

- **BlockPatterns**: previously, it contained code related to the WC core patterns registration and some actions related to AI patterns. After the refactor, it only has the necessary code to register patterns, both from WC core and from the PTK API (using the new `PatternRegistry` class).
Note: GitHub is hiding the diff for this class because it’s big, but please make sure to load it and review it as well.

- **PatternRegistry** (new class): handles the registration of patterns. This code was previously on `BlockPatterns`. It’s essentially the same code but adapted to register also patterns coming from the PTK API (not from files).

- **AIPatterns** (new class): handles everything related to the patterns with AI (there’s no new code here, it’s all extracted from `BlockPatterns`).

- **PTKPatternsStore** (new class): it handles fetching PTK patterns and caching them to avoid making lots of requests to the PTK API.

- **PTKClient** (new class): handles the requests to the PTK API to fetch patterns.

In case this helps, this is a more visual diagram to understand the structure:
Screenshot 2024-05-15 at 12 17 04

### How to test the changes in this Pull Request:

We are going to test with one pattern from each of the PTK categories:
- Intro: `Intro: Left-aligned header and button` pattern.
- About: `About: Fullwidth image, content pull right` pattern.
- Services: `Services: Three column pricing table` pattern.
- Testimonials: `Review: A quote with scattered images` pattern.

0. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com` and make sure `Allow usage of WooCommerce to be tracked ` is **unchecked**.
1. Create a new page or post and search for and try to insert the mentioned patterns, make sure they are **not** available.
2. Go to `wp-admin/tools.php?page=woocommerce-admin-test-helper`, go to `Features`, and **enable** `pattern-toolkit-full-composability`.
3. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com` and **check** the `Allow usage of WooCommerce to be tracked `.
4. Create a new page or post and insert the mentioned patterns. It should be available this time.
5. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com` and **uncheck** the `Allow usage of WooCommerce to be tracked `.
6. Create a new page or post and search for and try to insert the mentioned patterns, make sure it's **not** available.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
[CYS] Refactor the pattern registration and add patterns from the PTK API.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/changelog/47306-add-patterns-from-ptk [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
```
Comment (nefeline) [inline plugins/woocommerce/changelog/47306-add-patterns-from-ptk]: I think this is a major change, no? :)

**Review by nefeline — APPROVED** [review]
Exceptional work, @albarin ! I really appreciate your attention to the details on this implementation. Everything tested perfectly: the addition and removal of the transients for caching the requests are also spot on when it comes to plugin deactivation and updating the option for usage tracking on the database.

The fact that you have worked on adding tests is also fantastic! 👏 

Just left a suggestion regarding the significance of this change, apart from that, we are off to a great start :)

quick question (not a blocker): did you test if things are still working as expected with AI with those changes in place? It's not super important in this first moment because these changes are behind the feature flag & we are planning to change the way we inject content in patterns down the line, but asking anyways so we can keep this in mind :)

File: plugins/woocommerce/changelog/47306-add-patterns-from-ptk (hunk 1 above)
Comment (albarin) [reply plugins/woocommerce/changelog/47306-add-patterns-from-ptk]: Yep, you are right, changing it!

**Comment by albarin** [comment]
@nefeline thanks for the review!

> quick question (not a blocker): did you test if things are still working as expected with AI with those changes in place? It's not super important in this first moment because these changes are behind the feature flag & we are planning to change the way we inject content in patterns down the line, but asking anyways so we can keep this in min

I just tried and in WooExpress AI still works just fine. However, I noticed an issue with the `Product Collection` block not being registered. I'll investigate that now.
Screenshot 2024-05-17 at 11 29 56

**Comment by albarin** [comment]
@nefeline I think the `Product Collection` was just a glitch from some error building locally and uploading to WooExpress, not it's working fine. Would you mind giving it a try yourself? Just double-check 🙏

File: plugins/woocommerce/src/Blocks/Patterns/PatternRegistry.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+			return;
+		}
+
+		if ( \WP_Block_Patterns_Registry::get_instance()->is_registered( $pattern_data['slug'] ) ) {
+			return;
+		}
+
+		if ( isset( $pattern_data['featureFlag'] ) && '' !== $pattern_data['featureFlag'] && ! Features::is_enabled( $pattern_data['featureFlag'] ) ) {
+			return;
+		}
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/Patterns/PatternRegistry.php:60]: Just out of curiosity, can you give more details on the first two checks? 
Could a pattern have a feature flag value?

**Review by gigitux — APPROVED** [review]
LGTM! Thanks for your brilliant work! 🚀

I just left a comment, but just out of curiosity! 🚢

File: plugins/woocommerce/src/Blocks/Patterns/PatternRegistry.php (hunk 2 above)
Comment (albarin) [reply plugins/woocommerce/src/Blocks/Patterns/PatternRegistry.php:60]: Sure, it was added here https://github.com/woocommerce/woocommerce/pull/46856/

Comment (gigitux) [reply plugins/woocommerce/src/Blocks/Patterns/PatternRegistry.php:60]: TIL! Thanks!

**Comment by nefeline** [comment]
> Would you mind giving it a try yourself? Just double-check 🙏

I just did & confirmed things are working as expected, thanks for checking 🙏

**Pull request merged**